File: web/src/calibration.ts
/**
 * Client-side calibration utilities for multi-sample enrollment.
 */

/**
 * Calculate L1 (Manhattan) distance between two feature vectors.
 */
export function l1Distance(features1: number[], features2: number[]): number {
  if (features1.length !== features2.length) {
    throw new Error('Feature vectors must have same length');
  }
  return features1.reduce((sum, val, i) => sum + Math.abs(val - features2[i]), 0);
}

/**
 * Average multiple feature vectors to create a more stable template.
 */
export function averageFeatures(featureVectors: number[][]): number[] {
  if (featureVectors.length === 0) {
    throw new Error('At least one feature vector required');
  }
  
  const n = featureVectors.length;
  const length = featureVectors[0].length;
  
  // Ensure all vectors have same length
  for (const vec of featureVectors) {
    if (vec.length !== length) {
      throw new Error('All feature vectors must have same length');
    }
  }
  
  // Average each position
  const averaged: number[] = [];
  for (let i = 0; i < length; i++) {
    const avg = featureVectors.reduce((sum, vec) => sum + vec[i], 0) / n;
    averaged.push(Math.round(avg));
  }
  
  return averaged;
}

/**
 * Calculate adaptive threshold (tau) based on sample variance.
 * Uses mean + multiplier * stddev to set threshold.
 */
export function calculateAdaptiveTau(
  template: number[],
  samples: number[][],
  baseTau: number = 400,
  multiplier: number = 2.0,
): number {
  if (samples.length === 0) {
    return baseTau;
  }
  
  // Calculate distances from template to each sample
  const distances = samples.map(sample => l1Distance(template, sample));
  
  if (distances.length < 2) {
    // Not enough samples for variance, use base + small margin
    const meanDist = distances[0] || 0;
    return Math.max(baseTau, Math.round(meanDist * 1.5));
  }
  
  // Calculate mean and standard deviation
  const meanDist = distances.reduce((a, b) => a + b, 0) / distances.length;
  const variance = distances.reduce((sum, d) => sum + Math.pow(d - meanDist, 2), 0) / distances.length;
  const stddev = Math.sqrt(variance);
  
  // Adaptive threshold: mean + multiplier * stddev
  const adaptiveTau = Math.round(meanDist + multiplier * stddev);
  
  // Ensure it's at least baseTau
  return Math.max(baseTau, adaptiveTau);
}

/**
 * Calculate quality metrics for a template.
 */
export function calculateTemplateQuality(
  template: number[],
  samples: number[][],
): { meanDistance: number; stddevDistance: number; consistencyScore: number } {
  if (samples.length === 0) {
    return { meanDistance: 0, stddevDistance: 0, consistencyScore: 0 };
  }
  
  const distances = samples.map(sample => l1Distance(template, sample));
  const meanDist = distances.reduce((a, b) => a + b, 0) / distances.length;
  const variance = distances.reduce((sum, d) => sum + Math.pow(d - meanDist, 2), 0) / distances.length;
  const stddev = Math.sqrt(variance);
  
  // Consistency score: lower stddev = higher consistency (0-1 scale)
  const consistencyScore = Math.max(0, Math.min(1, 1 - (stddev / 200)));
  
  return {
    meanDistance: Math.round(meanDist * 100) / 100,
    stddevDistance: Math.round(stddev * 100) / 100,
    consistencyScore: Math.round(consistencyScore * 1000) / 1000,
  };
}

File: huproof/core/calibration.py
"""Template calibration and adaptive threshold calculation."""

from typing import Sequence

from .logging import get_logger

logger = get_logger()


def calculate_l1_distance(features1: Sequence[int], features2: Sequence[int]) -> int:
    """Calculate L1 (Manhattan) distance between two feature vectors.
    
    Parameters
    ----------
    features1: Sequence[int]
        First feature vector
    features2: Sequence[int]
        Second feature vector
        
    Returns
    -------
    int
        Sum of absolute differences
    """
    if len(features1) != len(features2):
        raise ValueError("Feature vectors must have same length")
    return sum(abs(a - b) for a, b in zip(features1, features2))


def average_features(feature_vectors: Sequence[Sequence[int]]) -> list[int]:
    """Average multiple feature vectors to create a more stable template.
    
    Parameters
    ----------
    feature_vectors: Sequence[Sequence[int]]
        List of feature vectors to average
        
    Returns
    -------
    list[int]
        Averaged feature vector (rounded to integers)
    """
    if not feature_vectors:
        raise ValueError("At least one feature vector required")
    
    n = len(feature_vectors)
    length = len(feature_vectors[0])
    
    # Ensure all vectors have same length
    for vec in feature_vectors:
        if len(vec) != length:
            raise ValueError("All feature vectors must have same length")
    
    # Average each position
    averaged = []
    for i in range(length):
        avg = sum(vec[i] for vec in feature_vectors) / n
        averaged.append(int(round(avg)))
    
    return averaged


def calculate_adaptive_tau(
    template: Sequence[int],
    samples: Sequence[Sequence[int]],
    base_tau: int = 400,
    multiplier: float = 2.0,
) -> int:
    """Calculate adaptive threshold (tau) based on sample variance.
    
    Uses mean + multiplier * stddev to set threshold that accommodates
    natural typing variation while rejecting impostors.
    
    Parameters
    ----------
    template: Sequence[int]
        Reference template vector
    samples: Sequence[Sequence[int]]
        Sample feature vectors from enrollment
    base_tau: int
        Minimum threshold (fallback if variance is too low)
    multiplier: float
        Standard deviation multiplier (higher = more lenient)
        
    Returns
    -------
    int
        Adaptive threshold value
    """
    if not samples:
        logger.warning("no_samples_for_tau", using_base=base_tau)
        return base_tau
    
    # Calculate distances from template to each sample
    distances = [calculate_l1_distance(template, sample) for sample in samples]
    
    if len(distances) < 2:
        # Not enough samples for variance, use base + small margin
        mean_dist = distances[0] if distances else 0
        return max(base_tau, int(mean_dist * 1.5))
    
    # Calculate mean and standard deviation
    mean_dist = sum(distances) / len(distances)
    variance = sum((d - mean_dist) ** 2 for d in distances) / len(distances)
    stddev = variance ** 0.5
    
    # Adaptive threshold: mean + multiplier * stddev
    adaptive_tau = int(mean_dist + multiplier * stddev)
    
    # Ensure it's at least base_tau
    tau = max(base_tau, adaptive_tau)
    
    logger.info(
        "adaptive_tau_calculated",
        mean=round(mean_dist, 2),
        stddev=round(stddev, 2),
        tau=tau,
        base_tau=base_tau,
    )
    
    return tau


def calculate_template_quality(
    template: Sequence[int],
    samples: Sequence[Sequence[int]],
) -> dict[str, float]:
    """Calculate quality metrics for a template.
    
    Returns metrics like consistency (lower variance = better) and
    average distance from template.
    
    Parameters
    ----------
    template: Sequence[int]
        Template vector
    samples: Sequence[Sequence[int]]
        Sample vectors
        
    Returns
    -------
    dict[str, float]
        Quality metrics: mean_distance, stddev_distance, consistency_score
    """
    if not samples:
        return {"mean_distance": 0.0, "stddev_distance": 0.0, "consistency_score": 0.0}
    
    distances = [calculate_l1_distance(template, sample) for sample in samples]
    mean_dist = sum(distances) / len(distances)
    variance = sum((d - mean_dist) ** 2 for d in distances) / len(distances)
    stddev = variance ** 0.5
    
    # Consistency score: lower stddev = higher consistency (0-1 scale, normalized)
    # Normalize by assuming max reasonable stddev is ~200
    consistency_score = max(0.0, min(1.0, 1.0 - (stddev / 200.0)))
    
    return {
        "mean_distance": round(mean_dist, 2),
        "stddev_distance": round(stddev, 2),
        "consistency_score": round(consistency_score, 3),
    }

